tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+    } else if (has_lf && has_crlf) {
+        fprintf(stderr, "Validation failed: %s contains mixed line endings (both LF and CRLF).\n", filepath.c_str());
+    } else if (has_lf) {
+        fprintf(stderr, "Validation failed: %s contains LF line endings.", filepath.c_str());
+
+        if (prev != LF) {
+            fprintf(stderr, " Also, it doesn't end with a newline.\n");
+        } else if (previous2 == LF) {
+            fprintf(stderr, " Also, it ends with multiple newlines.\n");
+        } else {
+            fprintf(stderr, "\n");
+        }
+    } else if (has_crlf) {
+        if (previous2 != CR || prev != LF) {
+            fprintf(stderr, "Validation failed: %s doesn't end with a newline.\n", filepath.c_str());
+        } else if (previous3 == LF) {
+            fprintf(stderr, "Validation failed: %s ends with multiple newlines.\n", filepath.c_str());
+        }
+    } else {
+        fprintf(stderr, "Validation failed: %s doesn't contain any newlines.\n", filepath.c_str());
+    }
+
+    if (has_utf8_bom) {
+        fprintf(stderr, "Validation failed: %s contains UTF-8 BOM characters.\n", filepath.c_str());
+    }
+
+    if (tab_characters != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu tab characters.\n", filepath.c_str(), tab_characters);
+    }
+
+    if (trailing_whitespace_lines != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu lines with trailing whitespace.\n", filepath.c_str(),
+            trailing_whitespace_lines);
+    }
+}
+
+int main() {
+    static constexpr array skipped_directories{
+        ".git"sv,
+        ".vs"sv,
+        ".vscode"sv,
+        "out"sv,
+        "vcpkg"sv,
+    };
+
+    static constexpr array skipped_extensions{
+        ".dll"sv,
+        ".exe"sv,
+        ".obj"sv,
+    };
+
+    // TRANSITION, P0202R3, use constexpr is_sorted()
+    assert(is_sorted(skipped_directories.begin(), skipped_directories.end()));
+    assert(is_sorted(skipped_extensions.begin(), skipped_extensions.end()));
+
+    vector<unsigned char> buffer; // reused for performance
+
+    try {
+        for (filesystem::recursive_directory_iterator rdi{"."}, last; rdi != last; ++rdi) {
+            if (!rdi->is_regular_file()) {
+                if (rdi->is_directory()) {
+                    const string filename = rdi->path().filename().string();
+                    if (binary_search(skipped_directories.begin(), skipped_directories.end(), filename)) {
+                        rdi.disable_recursion_pending();
+                    }
+                }
+
+                continue;
+            }
+
+            const string extension = rdi->path().extension().string();
+
+            if (binary_search(skipped_extensions.begin(), skipped_extensions.end(), extension)) {
+                continue;
+            }
+
+            const string filepath = rdi->path().string();
+
+            scan_file(filepath, buffer);
+        }
+    } catch (const exception& e) {

```

Comment:
should this be runtime_error instead of exception?

Reply:
`exception` catches more stuff. I could specifically catch `runtime_error` to print out `"runtime_error: meow"` if you want.

Reply:
I'll remove `try` / `catch` entirely, making `BinaryFile`'s ctor print to `stderr` for `fopen()` failure. Unexpected exceptions can then be debugged.